graphite-ch-optimizer.go

```diff
@@ -0,0 +1,310 @@
+		// See ClickHouse documentation for further options
+		"server-dsn": "tcp://localhost:9000?&optimize_throw_if_noop=1&read_timeout=3600&debug=true",
+		// Ignore partitions which were merged less than 3 days before
+		"optimize-interval": 60 * 60 * 24 * 3,
+	})
+	v.SetDefault("daemon", map[string]interface{}{
+		"one-shot":      false,
+		"loop-interval": 60 * 60,
+		"dry-run":       false,
+	})
+	v.SetDefault("logging", map[string]interface{}{
+		"output":    "-",
+		"log-level": "info",
+	})
+	defaultConfig := v.AllSettings()
+	log.Traceln(defaultConfig)
+	pflag.CommandLine.SortFlags = false
+	customConfig := pflag.StringP("config", "c", "", "Filename of the custom config. CLI arguments override it")
+	fc := pflag.NewFlagSet("clickhouse", 0)
+	fd := pflag.NewFlagSet("daemon", 0)
+	fl := pflag.NewFlagSet("logging", 0)
+	printDefaults := pflag.Bool("print-defaults", false, "Print default config values and exit")
+	fc.StringP("server-dsn", "s", v.GetString("clickhouse.server-dsn"), "DSN to connect to ClickHouse server")
+	fc.Uint("optimize-interval", v.GetUint("clickhouse.optimize-interval"), "The active partitions won't be optimized more than once per this interval, seconds")
+	fd.Bool("one-shot", v.GetBool("daemon.one-shot"), "Program will make only one optimization instead of working in the loop (true if dry-run)")
+	fd.Uint("loop-interval", v.GetUint("daemon.loop-interval"), "Daemon will check if there partitions to merge once per this interval, seconds")
+	fd.BoolP("dry-run", "n", v.GetBool("daemon.dry-run"), "Will print how many partitions would be merged without actions")
+	fl.String("output", v.GetString("logging.output"), "The logs file. '-' is accepted as STDOUT")
+	fl.String("log-level", v.GetString("logging.level"), "Valid options are: panic, fatal, error, warn, warning, info, debug, trace")
+	pflag.CommandLine.AddFlagSet(fc)
+	pflag.CommandLine.AddFlagSet(fd)
+	pflag.CommandLine.AddFlagSet(fl)
+	pflag.Parse()
+	v.SetConfigFile(*customConfig)
+	v.ReadInConfig()
+	log.Trace(v.ConfigFileUsed())
+	if *printDefaults {
+		t, err := toml.TreeFromMap(v.AllSettings())
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Println(t.String())
+		os.Exit(0)
+	}
+	fc.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("clickhouse."+f.Name, f)
+	})
+	fd.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("daemon."+f.Name, f)
+	})
+	fl.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("logging."+f.Name, f)
+	})
+	if v.GetBool("daemon.dry-run") {
+		v.Set("daemon.one-shot", true)
+	}
+	v.Unmarshal(c)
+	log.Traceln(c)
+	return c
+}
+
+func main() {
+	if cfg.Daemon.OneShot {
+		optimize(cfg)
+	} else {
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			log.Trace("Starting loop function")
+			for {
+				err := optimize(cfg)
+				if err != nil {
+					log.Errorf("Optimization failed: %s", err)
+				}
+				time.Sleep(time.Second * time.Duration(cfg.Daemon.LoopInterval))
+			}
+		}()
+		wg.Wait()
+	}
+}
+
+func optimize(cfg *Config) error {
```

Comment:
please make use of comments and some newlines in this method too. newlines are perfect for grouping stuff and for readability